File: Ahorn/entities/CustomTorch.jl
module XaphanHelperCustomTorch

using ..Ahorn, Maple

@mapdef Entity "XaphanHelper/CustomTorch" CustomTorch(x::Integer, y::Integer, width::Integer=16, height::Integer=16, playLitSound::Bool=false, startLit::Bool=false, flag::String="", color::String="ffa500", sprite::String="objects/XaphanHelper/CustomTorch/torch", alpha::Number=1.00, startFade::Integer=48, endFade::Integer=64, sound::String="event:/game/05_mirror_temple/torch_activate")

const placements = Ahorn.PlacementDict(
    "Custom Torch (Xaphan Helper)" => Ahorn.EntityPlacement(
        CustomTorch,
        "point",
        ),
    )

Ahorn.selection(entity::CustomTorch) = Ahorn.getEntityRectangle(entity)

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::CustomTorch, room::Maple.Room)
    sprite = get(entity.data, "sprite", "objects/XaphanHelper/CustomTorch/torch")
    Ahorn.drawSprite(ctx, "$(sprite)00.png", 8, 8)
end

end

File: Loenn/entities/CustomTorch.lua
local utils = require("utils")

local CustomTorch = {}

CustomTorch.name = "XaphanHelper/CustomTorch"
CustomTorch.depth = 2000
CustomTorch.fieldOrder = {
    "x", "y", "sprite", "sound", "startLit", "playLitSound", "color", "alpha", "startFade", "endFade", "flag"
}
CustomTorch.fieldInformation = {
    color = {
        fieldType = "color"
    },
    startFade = {
        fieldType = "integer",
    },
    endFade = {
        fieldType = "integer",
    }
}
CustomTorch.placements = {
    name = "CustomTorch",
    data = {
        playLitSound = false,
        startLit = false,
        flag = "",
        color = "FFA500",
        sprite = "objects/XaphanHelper/CustomTorch/torch",
        alpha = 1.00,
        startFade = 48,
        endFade = 64,
        sound = "event:/game/05_mirror_temple/torch_activate"
    }
}

function CustomTorch.texture(room, entity)
    local texture = entity.sprite or "objects/XaphanHelper/CustomTorch/torch"
  
    return texture .. "00"
end

function CustomTorch.selection(room, entity)
    return utils.rectangle(entity.x, entity.y , 16, 16)
end

CustomTorch.offset = {0, 0}

return CustomTorch